layer.md

```diff
@@ -166,31 +166,29 @@ The resulting tar archive for `rootfs-c9d-v1.s1` has the following entries:
 ./etc/.wh.my-app-config
 ```
 
-Where the basename name of `./etc/my-app-config` is now prefixed with `.wh.`, and will therefore be removed when the changeset is applied.
+where the basename name of `./etc/my-app-config` is now prefixed with `.wh.`, and will therefore be removed when the changeset is applied.
 
-## Applying
+## Applying Changesets
 
-Layer Changesets of [mediatype](./media-types.md) `application/vnd.oci.image.layer.tar+gzip` are applied rather than strictly extracted in normal fashion for tar archives.
+Layer Changesets of [mediatype](./media-types.md) `application/vnd.oci.image.layer.tar+gzip` are _applied_, rather than simply extracted as tar archives.
 
-Applying a layer changeset requires consideration for the [whiteout](#whiteouts) files.
-In the absence of any [whiteout](#whiteouts) files in a layer changeset, the archive is extracted like a regular tar archive.
+Applying a layer changeset requires special consideration for the [whiteout](#whiteouts) files.
 
+In the absence of any [whiteout](#whiteouts) files in a layer changeset, the archive is extracted like a regular tar archive.
 
 ### Changeset over existing files
 
-This section covers applying an entry in a layer changeset, if the file path already exists.
+This section describes applying an entry from a layer changeset if the file path already exists.
 
-If the file path is a directory, then the existing path just has it's attribute set from the layer changeset for that filepath.
-If the file path is any other file type (regular file, FIFO, etc), then the:
-* file path is unlinked (See [`unlink(2)`](http://linux.die.net/man/2/unlink))
-* create the file
-    * If a regular file then content written.
-* set attributes on the filepath
+If the entry and the existing path are both directories, then the existing path's attributes MUST be replaced by those of the entry in the changeset.
+In all other cases, the implementation MUST do the semantic equivalent of the following:
+- removing the file path (e.g. [`unlink(2)`](http://linux.die.net/man/2/unlink) on Linux systems)
```

Comment:
> I think we need to be clear that this is recursive only when the lstat (shallow) type of the resource is dir.

No, it's only *not* a recursive delete when both the existing entry (lstat) and tar entry are dirs.  See the implementation in opencontainers/image-tools#42 for machine-readable phrasing.

> If `/a/b/…` exists, a non-directory entry `/a/b` would cause removal, recursively.

Yes.

> Further entries prefixed with an already encountered, non-directory entry, would be ignored or cause an error.

Already-encountered-ness has nothing to do with it.  I think we should just unpack as we read through the tarball.  Only the current `.wh.*` handling [jumps the tar order][1], and the remove-before-unpacking behavior is for non-whiteout entries.

> If `/a/b/` exists and a new directory entries `/a/b/…` are encountered, the result will be a union of the two directory trees, unless whiteouts apply.

This is true even without the “directory” limit on `/a/b/…`.  If the filesystem has `/a/b/c` and an `/a/b/not-c` entry is found in the tarball, then the result will have both `/a/b/c` and `/a/b/not-c`.  If an `/a/b` entry is found in the tarball, then it's just a metadata clobber (the result will still have `/a/b/c`).

[1]: https://github.com/opencontainers/image-spec/blob/d4ca161f431b7598b2ca58d58eecea6fcd5dfabb/layer.md#whiteouts